radix-engine/src/wasm/wasmi.rs

```diff
@@ -11,279 +14,478 @@ use crate::wasm::constants::*;
+    }
+    Ok(data[ptr..ptr + len].to_vec())
+}
 
-                Ok(Some(RuntimeValue::I64(buffer.as_i64())))
-            }
-            WRITE_SUBSTATE_FUNCTION_ID => {
-                let handle = args.nth_checked::<u32>(0)?;
-                let data_ptr = args.nth_checked::<u32>(1)?;
-                let data_len = args.nth_checked::<u32>(2)?;
+fn write_memory(
+    mut store: impl AsContextMut,
+    memory: Memory,
+    ptr: u32,
+    data: &[u8],
+) -> Result<(), InvokeError<WasmRuntimeError>> {
+    let mut store_ctx = store.as_context_mut();
+    let mem_data = memory.data(&mut store_ctx);
+
+    if ptr as usize > mem_data.len() || ptr as usize + data.len() > mem_data.len() {
+        return Err(InvokeError::SelfError(WasmRuntimeError::MemoryAccessError));
+    }
 
-                self.runtime
-                    .write_substate(handle, self.read_memory(data_ptr, data_len)?)?;
+    memory
+        .write(&mut store.as_context_mut(), ptr as usize, data)
+        .or_else(|_| Err(InvokeError::SelfError(WasmRuntimeError::MemoryAccessError)))
+}
 
-                Ok(None)
-            }
-            UNLOCK_SUBSTATE_FUNCTION_ID => {
-                let handle = args.nth_checked::<u32>(0)?;
+fn read_slice(
+    store: impl AsContextMut,
+    memory: Memory,
+    v: Slice,
+) -> Result<Vec<u8>, InvokeError<WasmRuntimeError>> {
+    let ptr = v.ptr();
+    let len = v.len();
 
-                self.runtime.unlock_substate(handle)?;
+    read_memory(store, memory, ptr, len)
+}
 
-                Ok(None)
-            }
-            GET_ACTOR_FUNCTION_ID => {
-                let buffer = self.runtime.get_actor()?;
+impl WasmiInstance {
+    fn get_export_func(&mut self, name: &str) -> Result<Func, InvokeError<WasmRuntimeError>> {
+        self.instance
+            .get_export(self.store.as_context_mut(), name)
+            .and_then(Extern::into_func)
+            .ok_or_else(|| {
+                InvokeError::SelfError(WasmRuntimeError::UnknownWasmFunction(name.to_string()))
+            })
+    }
+}
 
-                Ok(Some(RuntimeValue::I64(buffer.as_i64())))
-            }
-            CONSUME_COST_UNITS_FUNCTION_ID => {
-                let n: u32 = args.nth_checked(0)?;
-                self.runtime
-                    .consume_cost_units(n)
-                    .map(|_| Option::None)
-                    .map_err(|e| e.into())
+impl HostError for InvokeError<WasmRuntimeError> {}
+
+impl From<Error> for InvokeError<WasmRuntimeError> {
+    fn from(err: Error) -> Self {
+        let e_str = format!("{:?}", err);
+        match err {
+            Error::Trap(trap) => {
+                let invoke_err = trap
+                    .downcast_ref::<InvokeError<WasmRuntimeError>>()
+                    .unwrap_or(&InvokeError::SelfError(
+                        WasmRuntimeError::InvalidExportReturn,
+                    ));
+                invoke_err.clone()
```

Comment:
It sucks that we have to clone the error. Would be nice if we can simply downcast and own the error.

Reply:
agree